### ISSUE TITLE
When metadata IDs wrap around, old objects get their text replaced in a way that feels like a bug

### ISSUE THREAD

**Issue by jasonrohrer** [jasonrohrer/OneLife#901]
Here's an ancient waystone that had its ID get reused after the wrap-around.

![image](https://github.com/jasonrohrer/OneLife/assets/12563681/95097e28-c617-4fb7-af32-2dc953148938)

On long-lived servers, this is going to happen, especially with all the different systems that use metadata now (writing, maps, photos).  There are only 16K metadata IDs before it wraps around and starts replacing the contents of the older ids.

It seems like there should be a way to put a UID inside the metadata itself that is somehow verifiable to detect that it's the correct data for the object in question.  Maybe the object ID that it's attached to, or something.

Also, in a duplication situation (like with photo negatives to positives, when the same metadata gets attached to multiple objects), what happens?  Is there a new metadata ID instantiated?  If so, then the metadata ID changes every time the object ID changes, which means it's safe to include the object ID in the metadata itself.

And it doesn't really matter if some note papers get mixed up in terms of content (if they have the same object ID).

What's really bad is when a waystone becomes a notepaper.

Oh.. and this is a little dumb that it doesn't already work this way... but why doesn't the metadata DB key on the full 32-bit ID, which includes the object ID?

So we can wrap-around the the metadata ID space without overwriting the metadata on other object types that use the same ID.

I.e.... if MID 1 is metadata for a permanent note, it would have a 32-bit ID of something like.... [1:3141]

Then later, if we wrap around in the MID space back to 1, and it's a pencil note, we'd have a 32-bit ID of [1:1615], which would be stored in a different record in the metadata database.

To automatically smooth the transition process, we could first look for the 32-bit version when looking up metadata, and if it's not found, go back to looking for the 14-bit version.  After all, the metadata DB stores 32-bit keys anyway.

**Comment by jasonrohrer** [comment]
After looking into the code here, this is too hard to fix.

New metadata IDs are NOT assigned when object change (for example, ink making a note permanent), and doing so would further pollute the metadata ID space more than it already is.

I also thought about devoting more than 14 bits to the metadata portion of a map cell ID (since we will probably never need 131K unique object IDs ever).  However, making that change would invalidate all the existing IDs in the map database.  Maybe it would make sense to do this after a map wipe, but currently, it wouldn't be great to wipe out all that metadata content.

So, I'm going to leave this alone for now, unless a new idea comes in the future about how to fix it without making things worse.

**Comment by jasonrohrer** [comment]
Actually, while falling asleep last night, I think I came up with a way to fix this.

**Comment by jasonrohrer** [comment]
Dammit... actually, what I came up with last night didn't end up working after all.

The most important thing here is that new metadata is prioritized over old.  When someone creates a new note or photograph or whatever, it must work.  This can be at the expense of ancient stuff getting overwritten, etc.  But preserving ancient stuff can never happen at the expense of newly-created stuff failing to work correctly.

The solution that I came up with last night had a bad side-effect of making newly created stuff occasionally pick up ancient messages.  This is very disturbing when it happens.  "I just created this note, and right away it says something different than what I wrote."

Having ancient messages get replaced over time is preferable, because who really remembers who made them or what they were supposed to say.
